util/winutil/svcdiag_windows.go

```diff
@@ -0,0 +1,297 @@
+	}
+	if ti.triggerSubtype != nil {
+		m.TriggerSubtype = ti.triggerSubtype.String()
+	}
+	return json.Marshal(m)
+}
+
+type _SERVICE_TRIGGER_INFO struct {
+	cTriggers uint32
+	pTriggers *_SERVICE_TRIGGER
+	_         *byte
+}
+
+func (sti *_SERVICE_TRIGGER_INFO) MarshalJSON() ([]byte, error) {
+	triggers := unsafe.Slice(sti.pTriggers, sti.cTriggers)
+	return json.Marshal(triggers)
+}
+
+// getSvcTriggerInfo obtains information about any system events that may be
+// used to start svc. Only relevant for demand-start (aka manual) services.
+func getSvcTriggerInfo(svc *mgr.Service) (*_SERVICE_TRIGGER_INFO, error) {
+	var desiredLen uint32
+	err := queryServiceConfig2(svc.Handle, windows.SERVICE_CONFIG_TRIGGER_INFO,
+		nil, 0, &desiredLen)
+	if err != windows.ERROR_INSUFFICIENT_BUFFER {
+		return nil, err
+	}
+
+	buf := make([]byte, desiredLen)
+	err = queryServiceConfig2(svc.Handle, windows.SERVICE_CONFIG_TRIGGER_INFO,
+		&buf[0], desiredLen, &desiredLen)
+	if err != nil {
+		return nil, err
+	}
+
+	return (*_SERVICE_TRIGGER_INFO)(unsafe.Pointer(&buf[0])), nil
+}
+
+// makeLogEntry consolidates relevant service information into a svcStateLogEntry.
+// We record the values of various service configuration constants as strings
+// so the the log entries are easy to interpret at a glance by humans.
+func makeLogEntry(svc *mgr.Service, status svc.Status, cfg mgr.Config) (entry svcStateLogEntry) {
+	entry.ServiceName = svc.Name
+
+	switch status.State {
+	case windows.SERVICE_STOPPED:
+		entry.State = "STOPPED"
+	case windows.SERVICE_START_PENDING:
+		entry.State = "START_PENDING"
+	case windows.SERVICE_STOP_PENDING:
+		entry.State = "STOP_PENDING"
+	case windows.SERVICE_RUNNING:
+		entry.State = "RUNNING"
+	case windows.SERVICE_CONTINUE_PENDING:
+		entry.State = "CONTINUE_PENDING"
+	case windows.SERVICE_PAUSE_PENDING:
+		entry.State = "PAUSE_PENDING"
+	case windows.SERVICE_PAUSED:
+		entry.State = "PAUSED"
+	case windows.SERVICE_NO_CHANGE:
+		entry.State = "NO_CHANGE"
+	default:
+		entry.State = fmt.Sprintf("Unknown constant %d", status.State)
+	}
+
+	switch cfg.ServiceType {
+	case windows.SERVICE_FILE_SYSTEM_DRIVER:
+		entry.ServiceType = "FILE_SYSTEM_DRIVER"
+	case windows.SERVICE_KERNEL_DRIVER:
+		entry.ServiceType = "KERNEL_DRIVER"
+	case windows.SERVICE_WIN32_OWN_PROCESS, windows.SERVICE_WIN32_SHARE_PROCESS:
+		entry.ServiceType = "WIN32"
+	default:
+		entry.ServiceType = fmt.Sprintf("Unknown constant %d", cfg.ServiceType)
+	}
+
+	switch cfg.StartType {
+	case windows.SERVICE_BOOT_START:
+		entry.StartupType = "BOOT_START"
+	case windows.SERVICE_SYSTEM_START:
+		entry.StartupType = "SYSTEM_START"
+	case windows.SERVICE_AUTO_START:
+		if cfg.DelayedAutoStart {
+			entry.StartupType = "DELAYED_AUTO_START"
+		} else {
+			entry.StartupType = "AUTO_START"
+		}
+	case windows.SERVICE_DEMAND_START:
+		entry.StartupType = "DEMAND_START"
+		if triggerInfo, err := getSvcTriggerInfo(svc); err == nil {
```

Comment:
Do we want an error field here?

Reply:
It's not super critical for our purposes, but it doesn't hurt. I'll add one.